src/main.rs

```diff
@@ -1,183 +1,153 @@
+                ..Default::default()
             }
+        )),
+        completion_provider: Some(CompletionOptions {
+            resolve_provider: Some(true),
+            ..Default::default()
+        }),
+        definition_provider: Some(true),
+        document_formatting_provider: Some(true),
+        rename_provider: Some(RenameProviderCapability::Simple(true)),
+        ..Default::default()
+    }).unwrap();
 
-            let length = length.ok_or_else(|| format_err!("missing Content-Length in request"))?;
+    if let Err(err) = connection.initialize(capabilities) {
+        error!("{:?}", err);
+        return Err(DUMMY_ERROR);
+    }
 
-            let mut body = vec![0; length];
-            stdin.read_exact(&mut body)?;
+    App {
+        files: HashMap::new(),
+        conn: connection,
+    }.main();
 
-            writeln!(self.log, "Raw: {:?}", std::str::from_utf8(&body).unwrap_or_default())?;
-            let req: Result<Request, _> = serde_json::from_slice(&body);
-            writeln!(self.log, "{:#?}", req)?;
+    if let Err(err) = io_threads.join() {
+        error!("Unable to shut down I/O threads: {}", err);
+        return Err(DUMMY_ERROR);
+    }
 
-            let req = match req {
-                Ok(req) => req,
-                Err(err) => {
-                    writeln!(self.log, "{:?}", err)?;
-                    self.send(&Response::error(None, err))?;
-                    continue;
-                }
-            };
+    Ok(())
+}
 
-            let id = req.id;
-            if let Err(err) = self.handle_request(req) {
-                writeln!(self.log, "{:?}", err)?;
-                self.send(&Response::error(id, err))?;
+struct App {
+    files: HashMap<Url, (AST, String)>,
+    conn: Connection,
+}
+impl App {
+    fn reply(&mut self, response: Response) {
+        trace!("Sending response: {:#?}", response);
+        self.conn.sender.send(Message::Response(response)).unwrap();
+    }
+    fn notify(&mut self, notification: Notification) {
+        trace!("Sending notification: {:#?}", notification);
+        self.conn.sender.send(Message::Notification(notification)).unwrap();
+    }
+    fn err<E>(&mut self, id: RequestId, err: E)
+        where E: std::fmt::Display
+    {
+        warn!("{}", err);
+        self.reply(Response::new_err(id, ErrorCode::UnknownErrorCode as i32, err.to_string()));
+    }
+    fn main(&mut self) {
+        while let Ok(msg) = self.conn.receiver.recv() {
+            trace!("Message: {:#?}", msg);
+            match msg {
+                Message::Request(req) => {
+                    let id = req.id.clone();
+                    match self.conn.handle_shutdown(&req) {
+                        Ok(true) => break,
+                        Ok(false) => if let Err(err) = self.handle_request(req) {
+                            self.err(id, err);
+                        },
+                        Err(err) => self.err(id, err),
```

Comment:
It's better to break out of the loop in this case: protocol error during shutdown still means that you should shut down

Reply:
Errors only happen inside `handle_shutdown` if it actually detected a shutdown request?

Reply:
Yup

Reply:
Specifically, if shutdown request was not followed by exit notification